### PR TITLE
Button styling improvements

### LINKED ISSUE #5735: Button styling: increase font size
**Context** 
When we implemented [M3 button styling](https://github.com/getodk/collect/issues/5567), we decided to slightly adjust the button height to make the hit target easier for folks. 

We noticed that the full-width button text looked drastically smaller as a result. When we looked into increasing the font size, it wasn't an easy fix because it was either too big or too small, and we would need to create something custom. 

**Soultion** 

- Increase button font size so it feels more unified 
- Remove bold font from inline buttons 

**Update Sept 25 [adding these ideas for us to pair on]**
- Make the black outline lighter (right now, it's visually distracting)
- Making the buttons scale nicely depending on your device. We might want to try scaling the buttons to the text size for tablets because the full-width buttons are too much. 

**Example of full-width inline button text being too small**

Image

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5735

#### Why is this the best possible solution? Were any other approaches considered?
Nothing to discuss here.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Things that I have changed:
- removed the bold font from buttons
- updated the color of borders in outline buttons

Please review all buttons to make sure they look as expected.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@alyblenkin please take a look at the changes and let me know what you think about the new font size. I used the same size as we already had in the main menu buttons as @lognaturel proposed.
In some cases it might be a problem for example:
Screenshot_1696924223
but it didn't look perfect with the smaller size too so we might need to think about doing something with those buttons or at last adding shorter labels maybe?

**Comment by alyblenkin** [comment]
@grzesiek2010 

First of all, removing the bold font makes it so much easier to read! 

I think the button font size for the menu works really well. However, in the form-filling experience, the increased font size of the buttons in comparison to the question font size looks too drastically different. And we want to try to avoid situations where the button label text goes on two lines because it almost doubles the width of the buttons, like in the example you shared.

It was worthwhile testing the larger font size as part of the new M3 button changes, but I don't think we should make them the same size as the menu buttons. We originally increased the height as part of some of the other button updates because a user brought up that the touch area is sometimes difficult for larger fingers. Unfortunately, increasing the height and font size so it is unified creates other issues, with the button labels being too long and taking up too much space on smaller devices. 

**Proposal:** To make the button and font size ratio feel more unified, I think we need to decrease the button height and font size back to what we had before. 

We also wanted to explore using a lighter grey for the outlined button. The dark outline is problematic with the "Back" and "Next" buttons in particular, and it can be distracting from the form. Maybe something like #C4C4C4 would be slightly lighter but still visible.

**Comment by grzesiek2010** [comment]
Proposal: To make the button and font size ratio feel more unified, I think we need to decrease the button height and font size back to what we had before.

Do you mean to what we have on master?

> We also wanted to explore using a lighter grey for the outlined button. The dark outline is problematic with the "Back" and "Next" buttons in particular, and it can be distracting from the form. Maybe something like #C4C4C4 would be slightly lighter but still visible.

I'm not sure... wouldn't they look like disabled buttons then?
Screenshot_1697032564

**Comment by alyblenkin** [comment]
> Do you mean to what we have on master?

Yes

> I'm not sure... wouldn't they look like disabled buttons then? 

Sorry for the slow reply! I don’t think they look like disabled buttons because of the blue text. However, #C4C4C4 doesn’t pass for accessibility in terms of contrast. I assume that a lot of companies, like Google, use a lighter outline but rely on text contrast.
I think a middle ground would likely be our best option at this point and go with a slightly darker grey #909090.  And we **don’t** want the bold font because it’s very difficult to read.

I also don’t want this work to eat too much into the other features we have planned for this release. So I think we should go forward with the slightly softer version, and we can always change it as we navigate the wonderful world of implementing M3 styling :)

**Comment by grzesiek2010** [comment]
What about the dark mode? What color should be used there?

**Comment by alyblenkin** [comment]
> What about the dark mode? What color should be used there?

Dark mode is fine! We don't want to make any changes there.

**Comment by grzesiek2010** [comment]
@alyblenkin according to https://github.com/material-components/material-components-android/blob/master/docs/components/Button.md#container-attributes-3 the `strokeColor` should be `?attr/colorOnSurface at 12% opacity` in our case it's `color_on_surface_low_emphasis` I used it and I think it looks ok. What do you think?

**Comment by alyblenkin** [comment]
> @alyblenkin according to https://github.com/material-components/material-components-android/blob/master/docs/components/Button.md#container-attributes-3 the `strokeColor` should be `?attr/colorOnSurface at 12% opacity` in our case it's `color_on_surface_low_emphasis` I used it and I think it looks ok. What do you think?

Ah perfect! I didn't know we had the Material colours implemented.

**Review by alyblenkin — APPROVED** [review]

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java [hunk 1]
```diff
@@ -416,4 +418,21 @@ public void widgetValueChanged() {
...
         }
     }
+
+    /*
+    Loop through each child view to identify buttons and dynamically adjust their font size based on
+    the current settings. This efficient approach eliminates the need to manually adjust button font
+    sizes for individual widgets. Furthermore, this method lays the groundwork for potential future
+    enhancements, such as extending font size adjustments to other view types like text views etc.
+     */
+    public void adjustButtonFontSize(ViewGroup view) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: This shouldn't need to be public right?

**Review by seadowg — CHANGES_REQUESTED** [review]
Looks good! I've left a comment about one small tweak I'd like.

One thing I was thinking: did you consider introducing a theme for widgets that overrides the default button style instead of the more "programmatic" approach (with `adjustFontSize`)? It might be worth experimenting with. The end result would be the same (that buttons would all have a different default style from the rest of the app), but it would still allow us to have `Button` views in widgets that set their own text style which the current approach would not allow.

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Fixed.

**Comment by grzesiek2010** [comment]
> One thing I was thinking: did you consider introducing a theme for widgets that overrides the default button style instead of the more "programmatic" approach (with adjustFontSize)?

`adjustFontSize` is used to set the font size dynamically based on the settings. A new them would be good if it was static right?

**Comment by seadowg** [comment]
> adjustFontSize is used to set the font size dynamically based on the settings. A new them would be good if it was static right?

Ah good point! We'd still need a programmatic part to update the theme which would probably be pretty messy.

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
If there is Cancel button (e.g. in settings) there isn't any outline and the button looks different than the other buttons in the app. Does this PR apply to all of buttons in the app?
Cancelbutton

**Comment by grzesiek2010** [comment]
> Does this PR apply to all of buttons in the app?

Yes if it's a button with a border (the outline buttons with white background) then the color should be updated across the app. If there were buttons without borders it's not in the scope of this pr. Should they have borders I don't know I think it's ok as it is.

**Comment by dbemke** [comment]
The form map icon/button has a darker outline than other buttons. Is this in the scope of this PR?

**Comment by grzesiek2010** [comment]
> The form map icon/button has a darker outline than other buttons. Is this in the scope of this PR?

image

this one?

**Comment by dbemke** [comment]
Yes. If it's in the scope of the PR there are also buttons with icons in forms with media
buttonswithicons

**Comment by grzesiek2010** [comment]
So yes those also should be updated. I've just fixed that so it should be fine now.

**Comment by alyblenkin** [comment]
> Yes if it's a button with a border (the outline buttons with white background) then the color should be updated across the app. If there were buttons without borders it's not in the scope of this pr. Should they have borders I don't know I think it's ok as it is.

It's a really good point @dbemke and it has caused a lot of confusion in the UX world! Cancel buttons are navigation links, so they don't need an outline. For example: 

**Navigation links (without an outline)**
- Cancel
- Do not delete 
- Close

You're not changing any information here, you are simply navigating back to the previous state.

**Action buttons (with outline)**
- Confirm 
- OK 
- Download 
- Create

The key here is the action button is changing the data or manipulating something.

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android 8.1, 10

**Verified cases:**
- buttons in the dark and light mode in the app

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android 13

**Pull request merged**